### PR TITLE
Add ETAS interface

### PR DESCRIPTION
**Pull request by felixn**
This patch adds support for ETAS CAN Interfaces, based on the ETAS BOA (Basic Open API)

It's ready from a functional point of view.
I'm not sure what else is required for merging - @felixdivo, what do you think?
I tried to follow the steps described in https://github.com/hardbyte/python-can/blob/develop/doc/development.rst#creating-a-new-interfacebackend as well as #1149

I've added a back2back test, which mostly passes - but it can obviously only run successfully if an actual hardware interface is available.
b2b
- test_timestamp_is_absolute fails because timestamps are HW timestamps relative to powerup, not relative to unix epoch. What's the expected behaviour?
- test_unique_message_instances fails because the API does not seem to allow creating multiple instances of one channel with differing settings regarding self-reception...

Are further tests, e.g. unit tests with mocked API/HW access, mandatory?

### CODE REVIEW

**Comment by felixdivo** [comment]
Hey @felixn, Felix here. ^^

At first glance, your PR looks very good! Mandatory is working code and proper documentation, of which you seem to have included both (I can't test the code since I do not have the required hardware). Further tests are not required.

Regarding the `timestamp`: According to [the docs](https://python-can.readthedocs.io/en/develop/message.html#can.Message.timestamp), they are relative to "epoch in seconds". For example, the `pcan` interface also translates timestamps like [this](https://github.com/hardbyte/python-can/blob/develop/can/interfaces/pcan/pcan.py#L20). Maybe one could move that piece of code to `can/util.py` and you can use it too? Could you do that?

Regarding the last test case: It's totally fine to skip the test case if that is not supported, just add this to your test case class (you can of course change the description as you see fit):

```python
def test_unique_message_instances(self):
    self.skipTest("Creating the test environment with additional buses is not supported")
```

**Comment by felixdivo** [comment]
If you choose to implement the suggested `timestamp` fix, you'd also need to [add a extension key to `setup.py`](https://github.com/hardbyte/python-can/blob/develop/setup.py#L35) and advice users to install the entire library with `pip install can[etas]` (or something like that).

**Comment by felixn** [comment]
Hi Felix ;)
Thanks for the quick feedback!

I've made changes you suggested to make the timestamps relative to epoch.
Luckily, I didn't need the additional uptime library, since in this case the timestamps are relative to a hardware timer.

Ready to merge from my point of view!

File: can/interfaces/etas/__init__.py [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+                f"OCI_GetCANControllerStatus failed with error 0x{ec:X}"
+            )
+        if status.stateCode & OCI_CAN_STATE_ACTIVE:
+            return can.BusState.ACTIVE
+        elif status.stateCode & OCI_CAN_STATE_PASSIVE:
+            return can.BusState.PASSIVE
+
+    @state.setter
+    def state(self, new_state: can.BusState) -> None:
+        # if new_state == can.BusState.ACTIVE:
```
Comment (felixdivo) [inline can/interfaces/etas/__init__.py:368]: If there is a problem with the disabled code, one should probably add a brief comment on what's up.

File: can/interfaces/etas/__init__.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+    def _detect_available_configs() -> List[can.typechecking.AutoDetectedConfig]:
+        nodeRange = CSI_NodeRange(CSI_NODE_MIN, CSI_NODE_MAX)
+        tree = ctypes.POINTER(CSI_Tree)()
+        ec = CSI_CreateProtocolTree(ctypes.c_char_p(b""), nodeRange, ctypes.byref(tree))
+        if ec != 0x0:
+            raise can.exceptions.CanOperationError(
+                f"CSI_CreateProtocolTree failed with error 0x{ec:X}"
+            )
+
+        nodes = []
```
Comment (felixdivo) [inline can/interfaces/etas/__init__.py:386]:
```suggestion
        nodes: Dict[str, str] = []
```

File: can/interfaces/etas/boa.py [hunk 3]
```diff
@@ -0,0 +1,672 @@
...
+    ("childrenProbed", ctypes.c_int),
+]
+
+CSI_CreateProtocolTree = _csi.CSI_CreateProtocolTree
+CSI_CreateProtocolTree.argtypes = [
+    ctypes.c_char_p,
+    CSI_NodeRange,
+    ctypes.POINTER(ctypes.POINTER(CSI_Tree)),
+]
+CSI_CreateProtocolTree.restype = BOA_ResultCode
```
Comment (felixdivo) [inline can/interfaces/etas/boa.py:105]: One could also add `CSI_CreateProtocolTree.errcheck = ...` to use [this](https://docs.python.org/3/library/ctypes.html#ctypes._FuncPtr.errcheck) (and similarly for the other methods). Just an idea, maybe it simplifies the bus class?

**Review by felixdivo — CHANGES_REQUESTED** [review]
Okay, I went through your code line-by-line (except for `can/interfaces/etas/boa.py`).

This is a well-written contribution! Thank you!

The changes I mentioned are all rather small (I hope), except for the `errcheck` idea. What do you think?

By the way: A final review by @hardbyte is also required, only then this will be merged.

File: can/interfaces/etas/__init__.py [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+from typing import Any, List, Optional, Tuple
+
+import can
+from .boa import *
+
+
+class EtasBus(can.BusABC):
+    def __init__(
+        self,
+        channel: Any,
```
[the remark this answers is not in the document]
Comment (felixn) [reply can/interfaces/etas/__init__.py]: :+1:

File: can/interfaces/etas/__init__.py [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+    ):
+        self.receive_own_messages = receive_own_messages
+
+        nodeRange = CSI_NodeRange(CSI_NODE_MIN, CSI_NODE_MAX)
+        self.tree = ctypes.POINTER(CSI_Tree)()
+        ec = CSI_CreateProtocolTree(
+            ctypes.c_char_p(b""), nodeRange, ctypes.byref(self.tree)
+        )
+        if ec != 0x0:
+            raise can.exceptions.CanInitializationError(
```
[the remark this answers is not in the document]
Comment (felixn) [reply can/interfaces/etas/__init__.py]: :+1:

File: can/interfaces/etas/__init__.py [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+
+        timerCapabilities = OCI_TimerCapabilities()
+        ec = OCI_GetTimerCapabilities(self.ctrl, ctypes.byref(timerCapabilities))
+        if ec != 0x0:
+            raise can.exceptions.CanInitializationError(
+                f"OCI_GetTimerCapabilities failed with error 0x{ec:X}"
+            )
+        self.tickFrequency = timerCapabilities.tickFrequency  # clock ticks per second
+
+        # all timestamps are hardware timestamps relative to powerup
```
[the remark this answers is not in the document]
Comment (felixn) [reply can/interfaces/etas/__init__.py]: :+1:

File: can/interfaces/etas/__init__.py [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+    ) -> Tuple[Optional[can.Message], bool]:
+        canMessages = (ctypes.POINTER(OCI_CANMessageEx) * 1)()
+
+        m = OCI_CANMessageEx()
+        canMessages[0].contents = m
+
+        count = ctypes.c_uint32()
+        remaining = ctypes.c_uint32()
+        if timeout is not None:
+            t = OCI_Time(int(timeout * self.tickFrequency))
```
[the remark this answers is not in the document]
Comment (felixn) [reply can/interfaces/etas/__init__.py]: :+1:

**Comment by felixn** [comment]
Thanks for the feedback @felixdivo - all valuable, I made the changes you requested.
@hardbyte - what do you think?

**Review by felixdivo — APPROVED** [review]
Nice!

**Comment by vincent-mailhol** [comment]
@nbusser, thanks for the mention. Indeed, Philipp Ahrendt and myself also implemented all the bindings for the ETAS interfaces. However, I never implemented the mock-up test thus my pull request failed the auto checks. Since then, I moved to some other projects. I do not think I will go back on this one day. None the less, what we did was tested with a wide variety of ETAS hardware (including the virtual bus) to make sure to capture all the edge cases.
@felixn thanks for your work. I would really be happy to see the ETAS interface merged to python CAN. I also recommend you to have a look at what we did. We did our best to cover all the use cases such as the older API (BOA version 1.0.0 to 1.4.0), the bus errors (Error active, passive, bus off…). We also added an interface for the optional bit timing parameters (time segment, Synchronization Jump Width…)

File: doc/interfaces/etas.rst [hunk 8]
```diff
@@ -0,0 +1,33 @@
+ETAS
+====
+
+This interface adds support for CAN interfaces by `ETAS`_.
+The ETAS BOA_ (Basic Open API) is used.
+Install the "ETAS ECU and Bus Interfaces – Distribution Package".
+Only Windows is supported.
```
Comment (vincent-mailhol) [inline doc/interfaces/etas.rst:7]: I propose to rephrase this line as below:

Only Windows is supported by this interface. The Linux kernel v5.13 (and greater) natively supports ETAS ES581.4, ES582.1 and ES584.1 USB modules. To use these under Linux, please refer to :ref:`socketcan`.

**Comment by felixn** [comment]
@vincent-mailhol & @nbusser :
Thanks for connecting the dots!
Still, I must say that this is a rather frustrating situation - discovering code in pull requests is next to impossible, and a lot of effort could have been saved...

I see several good ideas from your code that I could adopt - as you mentioned.
However, the code as it is right now is functional and a big improvement from "no support at all".
@hardbyte - are you ok with merging this as-is, and extending in a second step?
I can't justify spending the effort to extend the implementation for more use cases without some kind of assurance the code will be considered for merging.

**Comment by felixdivo** [comment]
> @hardbyte - are you ok with merging this as-is, and extending in a second step?

I'm not him, but I'd say it's fine to merge it like this. However, improving the documentation with the small [suggestion](https://github.com/hardbyte/python-can/pull/1144#discussion_r753292233) by @vincent-mailhol would still be useful in this PR. Afterwards, I'd personally say it's ready to be merged.

**Comment by hardbyte** [comment]
> > @hardbyte - are you ok with merging this as-is, and extending in a second step?
> 
> I'm not him, but I'd say it's fine to merge it like this. However, improving the documentation with the small [suggestion](https://github.com/hardbyte/python-can/pull/1144#discussion_r753292233) by @vincent-mailhol would still be useful in this PR. Afterwards, I'd personally say it's ready to be merged.

@felixn thanks for your contribution. I understand the frustration and uncertainty, I'm really disappointed when I see changes not make it in and also when efforts are duplicated. It's at least partly my fault for trying to stay across all core changes - despite the fact I haven't used CAN professionally in about 7 years(!)

I  try to ensure we only merge clear and maintainable code but I'm (obviously) not able to keep up. I plan to make a few tweaks over my Christmas break primarily to ensure things are more sustainable. Long story short, as long as a core contributor like @felixdivo approves a change it shouldn't need me.

**Review by hardbyte — APPROVED** [review]

**Comment by felixdivo** [comment]
@felixn I just resolved the conflicts. If you address above [comment](https://github.com/hardbyte/python-can/pull/1144/#discussion_r753292233), we can merge this contribution and include it in the next major release that will be coming out soon.

**Comment by felixn** [comment]
@hardbyte - thanks for the feedback! No worries, I completely understand. The time and motivation necessary for spare time open source projects is not always the same - at least, that's how it is for me.

@felixdivo - thanks for resolving the conflicts, and for proposing to include it in the next release. I made the suggested changes.
After that, I can spend some time on porting/including the additional functionality from @vincent-mailhol's PR.

File: doc/interfaces/etas.rst [hunk 9]
```diff
@@ -0,0 +1,33 @@
+ETAS
+====
+
+This interface adds support for CAN interfaces by `ETAS`_.
+The ETAS BOA_ (Basic Open API) is used.
+Install the "ETAS ECU and Bus Interfaces – Distribution Package".
+Only Windows is supported by this interface. The Linux kernel v5.13 (and greater) natively supports ETAS ES581.4, ES582.1 and ES584.1 USB modules. To use these under Linux, please refer to :ref:socketcan.
```
Comment (felixdivo) [inline doc/interfaces/etas.rst:7]:
```suggestion
Only Windows is supported by this interface.
The Linux kernel v5.13 (and greater) natively supports ETAS ES581.4, ES582.1 and ES584.1 USB modules.
To use these under Linux, please refer to :ref:`socketcan`.
```

**Review by felixdivo — APPROVED** [review]
I'm just gonna change this small thing :)

**Pull request merged**